components/ws-daemon/nsinsider/main.go

```diff
@@ -528,3 +562,17 @@ const (
 	// FlagAtRecursive: Apply to the entire subtree: https://elixir.bootlin.com/linux/latest/source/include/uapi/linux/fcntl.h#L112
 	flagAtRecursive = 0x8000
 )
+
+func addConnectionLimitRule(limit int) error {
+	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
+	defer cancel()
+	rule := fmt.Sprintf("add rule ip gitpod ratelimit ip protocol tcp ct state new meter connmeter { ip daddr & 0.0.0.0 timeout 1m limit rate over %d/minute } counter drop", limit)
+
+	// the go library does not support meters, so use nft instead
+	cmd := exec.CommandContext(ctx, "/usr/sbin/nft", rule)
```

Comment:
Since this function runs before `pivot_root`, I believe it is safe to specify the `nftable` bin with the complete path at this stage. Am I right? If yes, please add a comment that why it is safe to run.

Reply:
I'm concerned that the user may be able to override this bin with a custom image.

Reply:
cc: @mrzarquon 
We should install the nftables command, right? If yes, we have to change the image and notify the self-hosted team.

Reply:
We do not enter the mount namespace of the workspace, only the network namespace. The binary [is part of ws-daemon](https://github.com/gitpod-io/gitpod/blob/f4ef8c3457b3bd352554c3feb76cb8f1bcc373c1/components/ws-daemon/leeway.Dockerfile#L25).

Reply:
💯 Can I ask you to add a comment about that? Why? I think this is a very important security point that is difficult to see from the code.